Annotated FieldInfo instance is mutated and reused if using type aliases

### Checks

* [x ] I added a descriptive title to this issue
* [ x] I have searched (google, github) for similar issues and couldn't find anything
* [ x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

<!-- Sorry to sound so draconian, but every second saved replying to issues is time spend improving pydantic :-) -->

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.2
            pydantic compiled: False
                 install path: /tmp/tmp.pnuPT7tTpf/lib/python3.9/site-packages/pydantic
               python version: 3.9.6 (default, Jun 30 2021, 10:22:16)  [GCC 11.1.0]
                     platform: Linux-5.13.0-rc7-x86_64-with-glibc2.33
     optional deps. installed: ['devtools', 'typing-extensions']

```


Working version
```py
from pydantic import BaseModel, constr, Field
from typing import Annotated

class MyModel(BaseModel):
  a: Annotated[str, Field(max_length=3)] = ''
  b: Annotated[str, Field(max_length=3)]

MyModel(a="123", b='123', c='123')
MyModel(a="1234", b='1234', c='1234')  # ValidationError
```

Failing version:
```py
from pydantic import BaseModel, constr, Field
from typing import Annotated

StrField3 = Annotated[str, Field(max_length=3)]

class MyModel(BaseModel):  # ValueError: `Field` default cannot be set in `Annotated` for 'b'
  a: StrField3 = ''
  b: StrField3
```

The `.default` is set to `''` on the FieldInfo instance at [fields.py:404](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/fields.py#L404).

This can be fixed by copying the `FieldInfo` before mutating it and associating it with ModelField.
```
from copy import copy
...
class ModelField(Representation):
     ...
    @staticmethod
    def _get_field_info(
        field_name: str, annotation: Any, value: Any, config: Type['BaseConfig']
    ) -> Tuple[FieldInfo, Any]:
          ....
          if field_info is not None:
               field_info = copy(field_info)
               ....
```

This appears related to #3714.

@Jaakkonen - did you try running the unit tests after changing to a copy of the `FieldInfo` object? I got some failures so it might be that this mutation is required in other places.

```
FAILED tests/test_generics.py::test_replace_types - IndexError: tuple index out of range
FAILED tests/test_schema.py::test_unenforced_constraints_schema[kwargs9-type_9] - IndexError: tuple index out of range

Results (36.46s):
    2287 passed
       2 failed
         - tests/test_generics.py:761 test_replace_types
         - tests/test_schema.py:1520 test_unenforced_constraints_schema[kwargs9-type_9]
       7 skipped
make: *** [Makefile:63: test] Error 1
```

--- a/changes/4067-adriangb.md
+++ b/changes/4067-adriangb.md
@@ -1,0 +1,1 @@
+fix in-place modification of FieldInfo that made it impossible to use PEP 593 type aliases

--- a/changes/4067-adriangb.md
+++ b/changes/4067-adriangb.md
@@ -1,1 +1,1 @@
-fix in-place modification of FieldInfo that made it impossible to use PEP 593 type aliases
+Fix in-place modification of `FieldInfo` that caused problems with PEP 593 type aliases

--- a/pydantic/fields.py
+++ b/pydantic/fields.py
@@ -1,3 +1,4 @@
+import copy
 from collections import Counter as CollectionCounter, defaultdict, deque
 from collections.abc import Hashable as CollectionsHashable, Iterable as CollectionsIterable
 from typing import (
@@ -446,6 +447,7 @@ class ModelField(Representation):
                 raise ValueError(f'cannot specify multiple `Annotated` `Field`s for {field_name!r}')
             field_info = next(iter(field_infos), None)
             if field_info is not None:
+                field_info = copy.copy(field_info)
                 field_info.update_from_config(field_info_from_config)
                 if field_info.default is not Undefined:
                     raise ValueError(f'`Field` default cannot be set in `Annotated` for {field_name!r}')

--- a/tests/test_annotated.py
+++ b/tests/test_annotated.py
@@ -132,3 +132,18 @@ def test_config_field_info():
     assert Foo.schema(by_alias=True)['properties'] == {
         'a': {'title': 'A', 'description': 'descr', 'foobar': 'hello', 'type': 'integer'},
     }
+
+
+def test_annotated_alias() -> None:
+    # https://github.com/samuelcolvin/pydantic/issues/2971
+
+    StrAlias = Annotated[str, Field(max_length=3)]
+    IntAlias = Annotated[int, Field(default_factory=lambda: 2)]
+
+    class MyModel(BaseModel):
+        a: StrAlias = 'abc'
+        b: StrAlias
+        c: IntAlias
+        d: IntAlias
+
+    assert MyModel(b='def') == MyModel(a='abc', b='def', c=2, d=2)

--- a/tests/test_annotated.py
+++ b/tests/test_annotated.py
@@ -1,3 +1,5 @@
+from typing import List
+
 import pytest
 from typing_extensions import Annotated
 
@@ -140,10 +142,13 @@ def test_annotated_alias() -> None:
     StrAlias = Annotated[str, Field(max_length=3)]
     IntAlias = Annotated[int, Field(default_factory=lambda: 2)]
 
+    Nested = Annotated[List[StrAlias], Field(description='foo')]
+
     class MyModel(BaseModel):
         a: StrAlias = 'abc'
         b: StrAlias
         c: IntAlias
         d: IntAlias
+        e: Nested
 
-    assert MyModel(b='def') == MyModel(a='abc', b='def', c=2, d=2)
+    assert MyModel(b='def', e=['xyz']) == MyModel(a='abc', b='def', c=2, d=2, e=['xyz'])

